alacritty/src/renderer/rects.rs

```diff
@@ -190,3 +195,231 @@ impl RenderLines {
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $normalize:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        $normalize,
+                        size_of::<Vertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
+
+            // Position.
+            add_attr!(2, gl::FLOAT, gl::FALSE, f32);
+
+            // Color.
+            add_attr!(4, gl::UNSIGNED_BYTE, gl::TRUE, u8);
+
+            // Reset buffer bindings.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self { vao, vbo, program })
+    }
+
+    pub fn draw(&mut self, size_info: &SizeInfo, rects: Vec<RenderRect>) {
+        unsafe {
+            // Bind VAO to enable vertex attribute slots specified in new().
+            gl::BindVertexArray(self.vao);
+
+            // Bind VBO only once for buffer data upload only.
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(self.program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        // Build rect vertices vector.
+        let mut vertices = RectVertices::new(rects.len());
+        for rect in &rects {
+            vertices.add_rect(center_x, center_y, rect);
+        }
+
+        unsafe {
+            // Upload and render accumulated vertices.
+            vertices.draw();
+
+            // Disable program.
+            gl::UseProgram(0);
+
+            // Reset buffer bindings to nothing.
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+            gl::BindVertexArray(0);
+        }
+    }
+}
+
+/// Helper struct to hold transient vertices for rendering.
+struct RectVertices {
+    vertices: Vec<Vertex>,
+}
+
+impl RectVertices {
+    fn new(rects: usize) -> Self {
+        let mut vertices = Vec::new();
+        vertices.reserve(rects * 6);
+        Self { vertices }
+    }
+
+    fn add_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
+        let color = Rgba {
+            r: rect.color.r,
+            g: rect.color.g,
+            b: rect.color.b,
+            a: (rect.alpha * 255.) as u8,
```

Comment:
Could we pass it as not requiring normalization and then normalize in shader like we do with text rendering? Should be the same as for perf, right? This should simplify vertex attrib pointer creation, since everything won't require normalization.

Reply:
This would require multiplying `rect.color.{rgb}` by 255 instead. Normalization is the preferred way.

Reply:
I mean, it'll be something like that.

```diff
diff --git a/alacritty/res/rect.f.glsl b/alacritty/res/rect.f.glsl
index 945eaf2..3810e7c 100644
--- a/alacritty/res/rect.f.glsl
+++ b/alacritty/res/rect.f.glsl
@@ -6,5 +6,5 @@ out vec4 FragColor;

 void main()
 {
-    FragColor = color;
+    FragColor = color / 255.;
 }
diff --git a/alacritty/src/renderer/rects.rs b/alacritty/src/renderer/rects.rs
index b415ad1..546f9c0 100644
--- a/alacritty/src/renderer/rects.rs
+++ b/alacritty/src/renderer/rects.rs
@@ -284,7 +284,7 @@ impl RectRenderer {
             add_attr!(2, gl::FLOAT, gl::FALSE, f32);

             // Color.
-            add_attr!(4, gl::UNSIGNED_BYTE, gl::TRUE, u8);
+            add_attr!(4, gl::UNSIGNED_BYTE, gl::FALSE, u8);

             // Reset buffer bindings.
             gl::BindVertexArray(0);

```

Though, we decide to not use the macro, we'd be fine with asking for normalization.

Reply:
I think the no-macro version is much more readable than trying to hide all this stuff.